cpp/src/arrow/compute/kernel.h

```diff
@@ -109,6 +109,9 @@ struct ARROW_EXPORT TypeMatcher {
   /// \brief Return true if this matcher accepts the data type.
   virtual bool Matches(const DataType& type) const = 0;
 
+  /// \brief Return true if this matcher accepts the combination types
+  virtual bool Matches(const std::vector<TypeHolder>& types) const { return true; }
+
```

Comment:
This class is for matching a single matcher (`this`) against a single type.

I don't think it's reasonable to extend the type matching this way as it makes type checking Turing-complete (wouldn't be easy/possible to encode the typing rules in a provably-decidable type system).

I'm sorry for this back and forth because I was the one that said "type matching machinery". I still think this logic shouldn't go on the output type resolver so I have a more precise suggestion this time. Since functions already have a way of implementing custom type matching rules -- `DispatchBest` —— you can add this logic in https://github.com/apache/arrow/blob/9ee6ea701e20d1b47934f977d87811624061d597/cpp/src/arrow/compute/kernels/scalar_compare.cc#L343-L345

and return `return arrow::compute::detail::NoMatchingKernel(this, *types);` or a `Status::NotImplemented` with a message that describes that you can't compare two decimals with different scales.

Reply:
Ah, the root cause of the problem that this PR is trying to solve is actually CompareFunction with two different decimal scales in BindNonRecursive , it is impossible to enter the logic of `DispatchBest`.
https://github.com/apache/arrow/blob/9ee6ea701e20d1b47934f977d87811624061d597/cpp/src/arrow/compute/expression.cc#L556-L563

This is why we did type judgment in the output type Resolver before, so that we can return not-ok in the first `FinishBind` and enter `DispatchBest`.

`DispatchBest` does not need to make additional judgments, because it will cast according to the decimal rules, ensuring that different input scales are cast to the same scales according to `DecimalPromotion::kAdd`.